### ISSUE TITLE
Client side certificates for authentication

### ISSUE THREAD

**Issue by LaKing** [cockpit-project/cockpit#8429]
Since putting such a sensitive door to the web, where a brute-force or some other password guessing mechanism could lead to the attacker gaining root-level access is generally a bad practice, I would suggest implementing authentication with client-side certificates. That could be used optionally replacing or in addition to password protection.

If the server has client-side certificate authentication enabled, clients can add the certificate to their browser and identify themselves securely.

```openssl pkcs12``` can create a client side certificate. The server can then verify it against a CA certificate.

[shell script to make a client certificate](https://github.com/coolaj86/nodejs-ssl-trusted-peer-example/blob/master/make-root-ca-and-certificates.sh)

Is there a chance this could be implemented?

**Comment by stefwalter** [comment]
@LaKing ... Here's my opinions on this topic:

 * The use case for "client certificates in addition to system (password) authentication" should come first.
 * The above use case should be implemented using a light web server such as nginx
 * Most of the work will be: documenting this and/or making some scripts / drop in files for easy set up.

Does that make sense? Are you interested in working on it?

 [ ] Documentation would go into the ```doc/guide``` directory.
 [ ] Installable scripts and or any necessary drop in files would go into ```src/ws``` directory.
 [ ] Tests to make sure this functionality continues to work would go into ```test/verify```

As a follow up this work could be integrated with @petervo 's bastion container: https://github.com/cockpit-project/cockpit/tree/master/containers/bastion

**Comment by cockpitcert** [comment]
Client side certificate identification is a must in Cockpit web management interface!

**Comment by LaKing** [comment]
@stefwalter Yes, I'm interested in implementing it.

So, if I understand it right, you would you suggest implementing this with a reverse proxy?

**Comment by stefwalter** [comment]
@LaKing, Yes that's correct, a reverse proxy. Here's a wiki page to get started: https://github.com/cockpit-project/cockpit/wiki/Proxying-Cockpit-over-NGINX

**Comment by LaKing** [comment]
@stefwalter I understand your suggestion.

This would cover the use case of certificates in addition to password-based authentication, but only for the outside world behind the firewall. This use case can be achieved by any reverse proxy that can handle websockets. NGINX, HAproy .. etc. However, I think these are out of the scope of the core cockpit project itself and are rather additional system-architectural solutions. 

What we could consider is a native implementation. 

We could have another entry point for cockpit, for example on a different port, and running as a different service, with its own set of configs - requiring a client certificate straight at the connection. Since client certificates are issued per user, they already contain the identification, the username in the CN field. This can be used then, and depending on the security level, this can be sufficient to log in. 

This use case cannot be covered by a reverse proxy only configuration. The goal would be to secure cockpit the same way as when ssh is turned passwordless. If password authentication is disabled, brute force or password guessing algorithms have no surface for attacks, whether internally, nor externally.

Basically, sysadmins who want to close all attack surfaces, have no other option yet, than to disable cockpit.

**Comment by stefwalter** [comment]
@stefwalter I understand your suggestion.

> This would cover the use case of certificates in addition to password-based authentication, but only for the outside world behind the firewall. This use case can be achieved by any reverse proxy that can handle websockets. NGINX, HAproy .. etc. However, I think these are out of the scope of the core cockpit project itself and are rather additional system-architectural solutions.

I would consider this more secure, as NGINX has way more hardening than the Cockpit web server code.

> What we could consider is a native implementation.

> We could have another entry point for cockpit, for example on a different port, and running as a different service, with its own set of configs - requiring a client certificate straight at the connection. Since client certificates are issued per user, they already contain the identification, the username in the CN field. This can be used then, and depending on the security level, this can be sufficient to log in.

If you're set on doin it this way, then see the doc/authentication.md file to get you started with how this would work. It is completely possible to make pluggable authenticators for this use case. In fact you could combine this with the above NGINX approach by passing information to Cockpit about which certificate was successfully used by the client.

> This use case cannot be covered by a reverse proxy only configuration. The goal would be to secure cockpit the same way as when ssh is turned passwordless. If password authentication is disabled, brute force or password guessing algorithms have no surface for attacks, whether internally, nor externally.

> Basically, sysadmins who want to close all attack surfaces, have no other option yet, than to disable cockpit.

Up to you. But keep in mind that Cockpit can also be accessed over SSH, with a bastion host configuration. See the bastion container for examples of this.

**Comment by martinpitt** [comment]
We've been working on this for several months now, see PR #11421
